modules/gapi/include/opencv2/gapi/util/variant.hpp

```diff
@@ -233,9 +240,76 @@ namespace util
     template<typename T, typename... Types>
     const T& get(const util::variant<Types...> &v);
 
+    template<std::size_t Index, typename... Types>
+    typename util::type_list_element<Index, Types...>::type& get(util::variant<Types...> &v);
+
+    template<std::size_t Index, typename... Types>
+    const typename util::type_list_element<Index, Types...>::type& get(const util::variant<Types...> &v);
+
     template<typename T, typename... Types>
     bool holds_alternative(const util::variant<Types...> &v) noexcept;
 
+
+    // Visitor
+    namespace detail
+    {
+        struct visitor_interface {};
+
+        template<typename R>
+        struct visitor_return_type_deduction_helper
+        {
+            using return_type = R;
+
+            // to be used in Lambda return type deduction context only
+            template<typename T>
+            return_type operator() (T&&);
+        };
+    }
+
+    // Special purpose `static_visitor` can receive additional arguments
+    template<typename R, typename Impl>
+    struct static_visitor : public detail::visitor_interface,
+                            public detail::visitor_return_type_deduction_helper<R> {
+
+        using return_type = typename detail::visitor_return_type_deduction_helper<R>::return_type;
+        using detail::visitor_return_type_deduction_helper<R>::operator();
```

Comment:
What's the idea behind that?

Reply:
for getting return type in common case: Lambda + Class Visitor - and keep one interface invocation point: visit() only  - it requires to unify `return_type` deduction mechanism. For `Lambda`  it is possible to take type of `decltype(visitor(get<0>(var)))` but for `Class Visitor` there is no `operator()` in base case, because it provides `operator() (std::size_t index, ...)`

So `operator()` uses only for `Class Visitor` only for deduction return type in `visit()`

Reply:
Got it, but a comment in the code would help here a lot.

Reply:
sure, i'll add it